src/Api/Endpoints/Projects/Instrument.php

```diff
@@ -0,0 +1,130 @@
+ * @license  Loris license
+ * @link     https://github.com/aces/Loris
+ */
+namespace LORIS\Api\Endpoints\Projects;
+
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Message\ResponseInterface;
+use \LORIS\Api\Endpoint;
+
+/**
+ * A class for handling the /projects/$projectname/instruments/$instrumentname
+ * endpoint.
+ *
+ * @category API
+ * @package  Loris
+ * @author   Xavier Lecours Boucher <xavier.lecours@mcin.ca>
+ * @license  Loris license
+ * @link     https://github.com/aces/Loris
+ */
+class Instrument extends Endpoint implements \LORIS\Middleware\ETagCalculator
+{
+    /**
+     * A cache of the results of the endpoint, so that
+     * it doesn't need to be recalculated for the ETag and handler
+     */
+    protected $responseCache = array();
+
+    /**
+     * Return which methods are supported by this endpoint.
+     *
+     * @return array supported HTTP methods
+     */
+    protected function allowedMethods() : array
+    {
+        return ['GET'];
+    }
+
+    /**
+     * Versions of the LORIS API which are supported by this
+     * endpoint.
+     *
+     * @return array a list of supported API versions.
+     */
+    protected function supportedVersions() : array
+    {
+        return [
+                "v0.0.1",
+                "v0.0.2",
+                "v0.0.3-dev",
+               ];
+    }
+
+    /**
+     * Handles a request that starts with /projects/$projectname/candidates
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        // FIXME: Validate project based permissions.
+
+        $pathparts = $request->getAttribute('pathparts');
+        if (count($pathparts) > 1) {
+            return (new \LORIS\Http\Response())
+                ->withStatus(404);
+        }
+
+        return (new \LORIS\Http\Response())
+            ->withHeader("Content-Type", "application/json")
+            ->withBody(
+                new \LORIS\Http\StringStream(
+                    $this->_toJSON($request)
+                )
+            );
+    }
+
+    /**
+     * Create an array representation of this endpoint's reponse body
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return string
+     */
+    private function _toJSON(ServerRequestInterface $request): string
+    {
+        $instrumentname = $request->getAttribute('pathparts')[0];
+
+        if (!isset($this->responseCache[$instrumentname])) {
+            try {
+                $instrument = \NDB_BVL_Instrument::factory(
+                    $instrumentname,
+                    null,
+                    null,
+                    true
+                );
+            } catch (\Exception $e) {
+                return (new \LORIS\Http\Response())
+                    ->withStatus(400)
```

Comment:
shouldn't this be a 500?

Reply:
or perhaps a 404 since the most likely way to get an exception is with a non-existent instrument (in which case the catch type should probably be more explicit.)